pkg/apis/kubermatic/v1/external_cluster.go

```diff
@@ -64,6 +64,131 @@ type ExternalClusterSpec struct {
+
+	Token string `json:"token,omitempty"` // Token is used to authenticate with the DigitalOcean API.
+}
+
+// KubeOneOpenstackCloudSpec specifies access data to an OpenStack cloud.
+type KubeOneOpenstackCloudSpec struct {
+	CredentialsReference *providerconfig.GlobalSecretKeySelector `json:"credentialsReference,omitempty"`
+
+	AuthURL  string `json:"authURL,omitempty"`
+	Username string `json:"username,omitempty"`
+	Password string `json:"password,omitempty"`
+
+	// project, formally known as tenant.
+	Project string `json:"project,omitempty"`
+	// project id, formally known as tenantID.
+	ProjectID string `json:"projectID,omitempty"`
+
+	Domain string `json:"domain,omitempty"`
+}
+
+// KubeOneVSphereCloudSpec credentials represents a credential for accessing vSphere.
+type KubeOneVSphereCloudSpec struct {
+	CredentialsReference *providerconfig.GlobalSecretKeySelector `json:"credentialsReference,omitempty"`
+
+	Server   string `json:"server,omitempty"`
+	Username string `json:"username,omitempty"`
+	Password string `json:"password,omitempty"`
+}
+
+// KubeOnePacketCloudSpec specifies access data to a Packet cloud.
+type KubeOnePacketCloudSpec struct {
+	CredentialsReference *providerconfig.GlobalSecretKeySelector `json:"credentialsReference,omitempty"`
+
+	APIKey    string `json:"apiKey,omitempty"`
+	ProjectID string `json:"projectID,omitempty"`
+}
+
+// KubeOneHetznerCloudSpec specifies access data to hetzner cloud.
+type KubeOneHetznerCloudSpec struct {
+	CredentialsReference *providerconfig.GlobalSecretKeySelector `json:"credentialsReference,omitempty"`
+
+	// Token is used to authenticate with the Hetzner cloud API.
+	Token string `json:"token,omitempty"`
+}
+
+// KubeOneNutanixCSIConfig contains credentials and the endpoint for the Nutanix Prism Element to which the CSI driver connects.
+type KubeOneNutanixCSIConfig struct {
+
+	// Prism Element Username for csi driver
+	Username string `json:"username,omitempty"`
+
+	// Prism Element Password for csi driver
+	Password string `json:"password,omitempty"`
+
+	// Prism Element Endpoint to access Nutanix Prism Element for csi driver
+	Endpoint string `json:"endpoint"`
+}
+
+// KubeOneNutanixCloudSpec specifies the access data to Nutanix.
+type KubeOneNutanixCloudSpec struct {
+	CredentialsReference *providerconfig.GlobalSecretKeySelector `json:"credentialsReference,omitempty"`
+
+	// ClusterName is the Nutanix cluster that this user cluster will be deployed to.
+	ClusterName   string `json:"clusterName"`
+	AllowInsecure bool   `json:"allowInsecure,omitempty"`
+	ProxyURL      string `json:"proxyURL,omitempty"`
+
+	Username string `json:"username,omitempty"`
+	Password string `json:"password,omitempty"`
+	// The Nutanix API (Prism Central) endpoint
+	Endpoint string `json:"endpoint"`
+	// The Nutanix API (Prism Central) port
+	Port string `json:"port,omitempty"`
+
+	// KubeOneNutanixCSIConfig for csi driver that connects to a prism element
+	CSI *KubeOneNutanixCSIConfig `json:"csi,omitempty"`
```

Comment:
* Why is this a dedicated field/struct and not embedded directly into this one?
* Why is it a pointer?

Reply:
no specific reason,  it's how it is for KKP cluster https://github.com/kubermatic/kubermatic/blob/master/pkg/apis/kubermatic/v1/cluster.go#L884

Reply:
Those fields are required in KubeOne, so I would just embed them in the struct like other fields. In KKP, those fields are optional (for some specific reasons), so it's a different situation.

Reply:
makes sense, I will make the changes